codeworld-api/src/CodeWorld/Driver.hs

```diff
@@ -2540,12 +2552,41 @@ simulationControls w
     | playbackSpeed w == 0 = [PlayButton, StepButton, SpeedSlider]
     | otherwise = [PauseButton, SpeedSlider]
 
+debugSimulationControls :: Wrapped [w] -> [Control [w]]
+debugSimulationControls w
+    | lastInteractionTime w > 5 = []
+    | playbackSpeed w == 0 && not (null (tail (state w))) = [PlayButton, StepButton, SpeedSlider, UndoButton]
+    | playbackSpeed w == 0 = [PlayButton, StepButton, SpeedSlider]
+    | otherwise = [PauseButton, SpeedSlider]
+
 simulationOf simInitial simStep simDraw =
-    runPauseable initial simStep simulationControls simDraw
+    runInspect simulationControls initial simStep (\_ r -> r) simDraw
   where
     initial =
         Wrapped {state = simInitial, playbackSpeed = 1, lastInteractionTime = 1000, isDragging = False}
 
+debugSimulationOf simInitial simStep simDraw =
+    runInspect debugSimulationControls initial step (\_ r -> r) draw
+  where
+    initial =
+        Wrapped {state = [simInitial], playbackSpeed = 1, lastInteractionTime = 1000, isDragging = False}
+    step dt (x:xs) = simStep dt x : x : xs
+    draw (x:xs) = simDraw x
+
 trace msg x = unsafePerformIO $ do
     hPutStrLn stderr (T.unpack msg)
     return x
+
+debugInteractionOf baseInitial baseStep baseEvent baseDraw = runInspect debugSimulationControls initial step event draw 
+  where
+    initial =
+        Wrapped {state = [baseInitial], playbackSpeed = 1, lastInteractionTime = 1000, isDragging = False}
+    event e (x:xs) = case x `seq` x' `seq` reallyUnsafePtrEquality# x x' of
```

Comment:
Since this is non-trivial and repeated, it's probably worth a function.  Something like:

    prependIfChanged :: (a -> a) -> [a] -> [a]
    prependIfChanged f (x:xs) = case x `seq` x' `seq` ifTrue# (reallyUnsafePtrEquality# x x') of
        0# -> x' : x : xs
        _ -> x : xs
      where x' = f x

Then you have

    step dt = prependIfChanged (baseStep dt)
    event e = prependIfChanged (baseEvent e)

and you're done!